From: James Bates <james.h.bates@gmail.com>
Subject: efifb: prevent null-deref when iterating dmi_list
Git-Commit: 55aa42f2e690157e254a6a6989f5f4ac928b35c8
Patch-Mainline: v3.13
References: bnc#848055

    efifb: prevent null-deref when iterating dmi_list

    The dmi_list array is initialized using gnu designated initializers, and
    therefore may contain fewer explicitly defined entries as there are
    elements in it. This is because the enum above with M_xyz constants
    contains more items than the designated initializer. Those elements not
    explicitly initialized are implicitly set to 0.

    Now efifb_setup() loops through all these array elements, and performs
    a strcmp on each item. For non explicitly initialized elements this will
    be a null pointer:

    This patch swaps the check order in the if statement, thus checks first
    whether dmi_list[i].base is null.

    Signed-off-by: James Bates <james.h.bates@gmail.com>
    Signed-off-by: David Herrmann <dh.herrmann@gmail.com>
    Signed-off-by: Tomi Valkeinen <tomi.valkeinen@ti.com>
Signed-off-by: Oliver Neukum <oneukum@suse.de>
---
 drivers/video/efifb.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/video/efifb.c b/drivers/video/efifb.c
index d2f6bcd..1b681ec 100644
--- a/drivers/video/efifb.c
+++ b/drivers/video/efifb.c
@@ -310,8 +310,8 @@ static int __init efifb_setup(char *options)
 		if (!*this_opt) continue;
 
 		for (i = 0; i < M_UNKNOWN; i++) {
-			if (!strcmp(this_opt, dmi_list[i].optname) &&
-					dmi_list[i].base != 0) {
+			if (dmi_list[i].base != 0 &&
+			!strcmp(this_opt, dmi_list[i].optname)) {
 				screen_info.lfb_base = dmi_list[i].base;
 				screen_info.lfb_linelength = dmi_list[i].stride;
 				screen_info.lfb_width = dmi_list[i].width;
-- 
1.8.4

